internal/statesync/reactor.go

```diff
@@ -800,24 +800,20 @@ func (r *Reactor) handleMessage(ctx context.Context, chID p2p.ChannelID, envelop
 // the respective channel. When the reactor is stopped, we will catch the signal
 // and close the p2p Channel gracefully.
 func (r *Reactor) processCh(ctx context.Context, ch *p2p.Channel, chName string) {
-	for {
-		select {
-		case <-ctx.Done():
-			r.logger.Debug("channel closed", "channel", chName)
-			return
-		case envelope := <-ch.In:
-			if err := r.handleMessage(ctx, ch.ID, envelope); err != nil {
-				r.logger.Error("failed to process message",
-					"err", err,
-					"channel", chName,
-					"ch_id", ch.ID,
-					"envelope", envelope)
-				if serr := ch.SendError(ctx, p2p.PeerError{
-					NodeID: envelope.From,
-					Err:    err,
-				}); serr != nil {
-					return
-				}
+	iter := ch.Receive(ctx)
```

Comment:
Separately from this PR—and possibly not worthwhile given libp2p.

All these process-channel methods seem to have such similar structure, and the variables seem to be the channel, the logger, and the callback to handle the envelope. I wonder if we could replace many/all of them with a top-level function.

```go
go processChannel(ctx, ch, logger, callback)
```

Even this one—which is slightly different from the others in that it takes the channel name—might be close enough to consolidate.

Anyway, just something for the back burner.

Reply:
Yeah, I would definitely prefer the callback approach over the current approach of having a select statement over the ChannelID. That seems like we are re-inventing a callback mechanism anyway.

Reply:
I've been thinking about this, having watched myself do the same thing a bunch... all of the reactors are structured similarly, so there's definitely some case for making things a bit simpler and sharing some infrastructure could be useful. 

I think I'm generally opposed to doing that refactor in this refactor, just because I want to think a bit about what the implications are for concurency within the reactor... and also do we even need to process the channels in different threads? Maybe, probably? I also want to think a bit about how this changes before we get to libp2p, and sort of iterating on on this maybe as part of a second step.

Reply:
I completely agree this is not something to include in this PR.